Top n accuracy metric

jwjohnson314 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
Fixes #10144


#### What does this implement/fix? Explain your changes.
This implements a top-n accuracy classification metric, for use with probabilistic class predictions in multiclass classification settings. A prediction is considered top-n accurate if the correct class is one of the n classes with the highest predicted probabilities. Probability maybe is too strong a term here, as all that is really required is a ranking of the predicted classes. 


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
ENH: added top_n_accuracy_score

Commit:
TST: added tests for top_n_accuracy_score

jnothman commented:
Thanks for the PR. There's quite a backlog of reviewing to do, so please be patient!

rth commented:
@jwjohnson314 Currently the tests you added fail on Pyhon 2.7, see for instance the [Travis CI output](https://travis-ci.org/scikit-learn/scikit-learn/jobs/195339639) as well the [style checking with flake8](https://travis-ci.org/scikit-learn/scikit-learn/jobs/195339643) so these issues would probably need to be addressed first.

jwjohnson314 committed:
fixed PEP8 style issues

jwjohnson314 commented:
Thanks @rth. I just realized that the numpy function I use in this metric was introduced in numpy 1.8, so it's not going to work as written with python 2.7. I'll have to revise it.

jwjohnson314 committed:
fixed PEP8 issues

jwjohnson314 committed:
fixed example in top_n_accuracy_score

lesteve commented:
The CircleCI error looks very similar to #8209. Try rebasing on master to see whether the CircleCI error goes away.

jwjohnson314 commented:
Thanks @lesteve. Maybe I screwed up my rebase?

rth commented:
That's rather strange, for instance this [Travis CI build](https://travis-ci.org/scikit-learn/scikit-learn/jobs/198049257) is marked as failed, but all the tests pass and I don't see any error messages there.

Circle CI is now failing when it cannot download the 20 newsgoups dataset (which was not downloaded before), so it is not related to this PR, the rebase looks fine..

jwjohnson314 commented:
Okay, I'm not that familiar with rebasing so wasn't sure. Let me know if I can do anything else.

jnothman commented:
This seems straightforward as an idea, but could you please provide a reference or two for which research community uses it?

jwjohnson314 commented:
Sure, the Imagenet challenge is one place where a top-5 and a top-1 metric are both used (image classification challenge): [ImageNet](http://link.springer.com/article/10.1007/s11263-015-0816-y?sa_campaign=email/event/articleAuthor/onlineFirst#)

lesteve committed:
Improve error message in plot_stock_market.py

qinhanmin2014 committed:
[MRG+1] Fix warnings in lgtm.com (remove redundant code) (#9719)

lesteve committed:
DOC fix misleading note about sphinx version

vrishank97 committed:
[MRG+1] Added exchange names to tickers in plot_stock_market.py (#9750)

lesteve committed:
Better treatment of empty data in plot_stock_market.py

Also use date rather than datetime because we do not need hour of day

taylorkm committed:
[MRG+1] Fix #9743: Adding parameter information to docstring. (#9757)

* Adding parameter information to docstring.

* Removing trailing whitespace from lines.

* Adding details of parameter to formal Parameters section.

* Shortened lines to meet requirements.

ktisha committed:
DOC: fix docstring of Imputer.fit (#9769)

lesteve committed:
Add expected parameter to check min_date, max_date and len_data

rather than hardcoding the logic in the quotes_historical_google function.

Some minor variable renaming.

brettkoonce committed:
various minor spelling tweaks (#9783)

jnothman committed:
FIX max date should use max, not min

jnothman committed:
FIX? Use ISO8601 dates and resolved URL for Google Finance

jnothman committed:
[MRG] MAINT allow deprecated functions to be pickled (#9787)

lesteve committed:
plot_stock_market.py checks are based on dates rather than on hard-coded values

venthur committed:
[MRG+1] MAINT Replace assert_array_equal with -assert_array_almost_equal where necessary. (#9774)

Mottl committed:
[MRG+2] add `var_smoothing` parameter to GaussianNB (#9681)

qinhanmin2014 committed:
hange alpha in plot_label_propagation_structure.py (#9788)

ogrisel committed:
FIX fmin_cobyla: iprint is deprecated, use disp (#9793)

lesteve committed:
Use CYTHON_VERSION on ubuntu build (#9797)

lesteve committed:
TRAVIS use cython dev on scipy-dev build (#9803)

jnothman committed:
TST Improve SelectFromModel tests (#9733)

Should fix one of the issues in #9393

albertcthomas committed:
FIX docstring of negative_outlier_factor_ in LOF (#9809)

osaid-r committed:
[MRG+1] remove 'matching' metric from docstrings (#9727)

scipy.spatial.distance.matching has been equivalent to hamming in scipy for a while.

Commit:
[MRG + 1] Fix ValueError in LabelEncoder when using inverse_transform on unseen labels (#9816)

ogrisel committed:
[MRG+1] Make TSNE trustworthiness test more robust (#9808)

Platform specific rounding errors can make the t-SNE algorithm
converge to varying quality results especially on small datasets
as done in this test.

We therefore need a lower threshold to account for that
variability.

lesteve committed:
Fix plot_out_of_core_classification.py. (#9815)

Starting from empty ~/scikit_learn_data got AttributeError: module
'sklearn.externals.six.moves.urllib_request' has no attribute
'urlretrieve'.

agitter committed:
DOC Add average precision definitions and cross references (#9583)

wdevazelhes committed:
DOC Fix error in documentation of trustworthiness (#9800)

1cph93 committed:
DOC Resolve typo in nearest neighbors regression docs (#9831)

albertcthomas committed:
[MRG+2] Clean common tests (#9340)

* rm dupes

* add check_supervised_y_no_nan in classifier checks: this implies changes for Ridge classifiers

* fix docstrings/comments

* FIX check fitting 1d X array raises error and FIX check fitting 2d array with only 1 feature either works or returns informative message

* modify check_fit2d_1sample in common tests so that it checks fitting either works or returns an informative message

* rm SpectralClustering case for the moment

* uniformize error messages for 1 sample case and fix SpectralClustering with ensure_min_samples=2

* add unit test for mean_shift when n_samples * quantile < 1

* FIX travis with ensure_min_samples=2 in _PLS

* try fix for failing tSNE test

* typos

* take @agramfort's review into account

* sc to fix string in gaussian_process

* add the class that is present to preserve information of previous message in gpc.py

vrishank97 committed:
ENH avoid FutureWarning in BaseSGD.set_params (#9802)

jnothman committed:
MAINT remove entire directory in make clean

qinhanmin2014 committed:
[MRG+1] Fix floating bug in roc_auc_score (#9786)

* ensure fpr and tpr are increasing in roc_curve with non integer sample weights
* add tests and move roc_auc_score from METRIC_UNDEFINED_BINARY to METRIC_UNDEFINED_MULTICLASS

lesteve committed:
FIX do not update conda as a temporary work-around for conda issue

https://github.com/conda/conda/issues/6030

lesteve committed:
MAINT explain the reason for conftest.py in the root folder

lesteve committed:
FIX test broken in numpy 1.14.dev due to array str changes

lesteve committed:
[MRG+1] Travis: move scipy-dev-wheels build to a cron job (#9852)

lesteve committed:
Fix test class to be runnable by pytest (#9860)

Test class with __init__ is not run by pytest

bucketclan committed:
DOC Removed a duplicate occurrence of a word in 'sklearn.neighbors.KNeighborsRegressor' docs (#9862)

* Removed a duplicate occurrence of the word 'but' from the 'Warning' section.

stevendbrown committed:
[MRG+1] Reduce runtime of graph_lasso (#9858)

* reduce runtime of graph_lasso

* fixed line length overrun

* added comment explaining the change

* changed explanation comment

oliblum90 committed:
[MRG + 1] enable metric = 'cosine' for tsne computation (#9623)

artiemq committed:
ENH Add verbose level into the RFE at the end of RFECV (#9848)

massich committed:
FIX PermissionError in datasets fetchers on Windows (#9847)

jnothman committed:
DOC fix 0.19 release date

jschendel committed:
DOC: Use setattr(self, ...) instead of self.setattr(...) (#9866)

syonekura committed:
[MRG+1] Setting max_iter/tol explicitly for SGD estimators in docs (#9776)

lesteve committed:
MAINT remove temporary conda work-around

Reverts 8de18e67b.

naoyak committed:
TRAVIS update packages to latest Anaconda 5.0.0 (#9871)

nielsenmarkus11 committed:
[MRG+1] Raise error when SparseSeries is passed into classification metrics (#7373)

* Raise error when SparseSeries is passed into roc_curve

* Changed "y_true" in second if block to "y_score"

* Remove code to import pandas and add sparseseries check to 'type_of_target' function. Finally, add 'type_of_target' call to _binary_clf_curve

* Remove pandas import and old comparison in roc_curve.

* Add test for 'type_of_target' function

* Add white space after commas

* Correct other white space issues

* Move type_of_target test into try clause, remove test_precision_recall_curve_pos_label since as multiclass it doesn't make sense

* Add test_precision_recall_curve_pos_label back in and also add test_binary_clf_curve to test new logic in _binary_clf_curve function

* Correct syntax and formatting.

* Remove trailing white space

* Correct validation logic

* Update test_multiclass.py per @jnothman 's request.

* Import SkipTest function.

* Remove extra white space from line 303

Commit:
[MRG+1] Fix typos in documentation (#9878)

* Fix grammatical error in

* Correct capitalization of "GitHub"

Used command `find . -type f -exec sed -i 's/Github/GitHub/g' {} \;` (h/t: https://stackoverflow.com/a/15402972)

hristog committed:
Remove unused variable alphas from the LARS example. (#9882)

GKjohns committed:
[MRG+1] Ledoit-Wolf behavior explanation (#9500)

* DOC add explaination of unexpected behavior to ledoit-wolf functions and class

* DOC add explaination of unexpected ledoit-wolf behavior to module documentation

* fix line that's longer than 80 chars, pep8 issue

* fix documentation changes to Ledoit_Wolf behavior explaination

* change bahavior explanation to a note in documentation

* remove unexpected behavior explanation from docstrings

* fix broken links in docs

qinhanmin2014 committed:
[MRG+1] BUG Avoid unexpected error in PCA when n_components='mle' (#9886)

* n_components mle

* update doc

* improve

* update what's new

* update what's new

amueller committed:
[MRG+1] check that splitters handle 2d y and give reasonable errors on multilabel y (#9744)

goncalo-rodrigues committed:
FIX Error in manifold.t_sne._kl_divergence for n_components > 2 (#9712)

Fixes #9711

Commit:
DOC fix a typo (#9892)

jnothman committed:
FIX missing return in deprecated function

jnothman committed:
CI avoid matplotlib 2.1.0

Fixes #9896

s4chin committed:
Fix 9865: Change code and add test (#9890)

lesteve committed:
Fix example for matplotlib 2.1 change. (#9897)

naoyak committed:
MAINT: remove deprecated sphinx config variables (#8828)

qinhanmin2014 committed:
[MRG+1] Improve the error message for some metrics when the shape of sample_weight is inappropriate (#9903)

qinhanmin2014 committed:
[MRG+1] Completely support binary y_true in roc_auc_score (#9828)

lesteve committed:
DOC: use intersphinx for links in gallery examples (#9909)

PGryllos committed:
[MRG+2] Use NearestNeighbors to speed up trustworthiness (#9861)

albertcthomas committed:
improve MinCovDet error when covariance of support data is 0 (#9910)

jnothman committed:
[MRG+1] DOC fix up news in master (#9899)

Commit:
DOC show plot and fix comments (#9925)

thechargedneutron committed:
[MRG] Deprecates gaussian process regression_models and correlation_models. (#9717)

Forgotten from earlier deprecation

vrishank97 committed:
[MRG+1] Update docstrings of KMeans.inertia_ (#9920)

[MRG+2] Update docstrings of KMeans.inertia_

jnothman committed:
[MRG] FIX Avoid accumulating forest predictions in non-threadsafe manner (#9830)

amueller committed:
DOC add review guidelines, make other ways to contribute more prominent and rephrase (#9745)

jkleint committed:
DOC Clarify docs for `make_classification` (#9918)

jnothman committed:
PEP8

jnothman committed:
[MRG] FIX Revert the addition of ndcg_score and dcg_score (#9932)

qinhanmin2014 committed:
[MRG+2] Deprecate reorder parameter in auc (#9851)

cbrummitt committed:
[MRG+1] Fix bug in StratifiedShuffleSplit for multi-label data with targets having > 1000 labels (#9922)

* Use ' '.join(row) for multi-label targets in StratifiedShuffleSplit because str(row) uses an ellipsis when len(row) > 1000
* Add a new test for multilabel problems with more than a thousand labels

lesteve committed:
TRAVIS test pandas dev version in scipy-dev build (#9940)

dalmia committed:
[MRG+1] Adding support for balanced accuracy (#8066)

* add function computing balanced accuracy

* documentation for the balanced_accuracy_score

* apply common tests to balanced_accuracy_score

* constrained to binary classification problems only

* add balanced_accuracy_score for CLF test

* add scorer for balanced_accuracy

* reorder the place of importing balanced_accuracy_score to be consistent with others

* eliminate an accidentally added non-ascii character

* remove balanced_accuracy_score from METRICS_WITH_LABELS

* eliminate all non-ascii charaters in the doc of balanced_accuracy_score

* fix doctest for nonexistent scoring function

* fix documentation, clarify linkages to recall and auc

* FIX: added changes as per last review See #6752, fixes #6747

* FIX: fix typo

* FIX: remove flake8 errors

* DOC: merge fixes

* DOC: remove unwanted files

* DOC update what's new

airalcorn2 committed:
[MRG+1] Add norm parameter to ComplementNB. (#9916)

josephsalmon committed:
Adding objective function in Ridge regression docstring (#9942)

Commit:
DOC fix inconsistency with current implementation (#9946)

jnothman committed:
Add DeprecationDict for #9677

thechargedneutron committed:
[MRG+1] DEPREC Change default for return_train_score to False (#9677)

jnothman committed:
DOC Correct deprecation version

albertcthomas committed:
[MRG+1] test that clustering returns int (#9912)

amueller committed:
FIX _BaseComposition._set_params with nested parameters (#9945)

qinhanmin2014 committed:
DOC Add missing class (#9955)

reiinakano committed:
[MRG+1] Fix cross_val_predict behavior for binary classification in decision_function (Fixes #9589) (#9593)

* fix cross_val_predict for binary classification in decision_function

* Add unit tests

* Add unit tests

* Add unit tests

* better fix

* fix conflict

* fix broken

* only calculate n_classes if one of 'decision_function', 'predict_proba', 'predict_log_proba'

* add test for SVC ovo in cross_val_predict

* flake8 fix

* fix case of ovo and imbalanced folds for binary classification

* change assert_raises to assert_raise_message for ovo case

* fix flake8 linetoo long

* add comments and clearer tests

* improve comments and error message for OvO

* fix .format error with L

* use assert_raises_regex for better error message

* raise error in decision_function special cases. change predict_log_proba missing classes to minimum numpy value

* fix broken tests due to special cases of decision_function

* add modified test for decision_function behavior that does not trigger edge cases

* fix typos

* fix typos

* escape regex .

* escape regex .

* address comments. one unaddressed comment

* simplify code

* flake

* wrong classes range

* address comments. adjust error message

* add warning

* change warning to runtimewarning

* add test for the warning

* Use assert_warns_message rather than assert_warns

Other minor fixes

* Note on class-absent replacement values

* Improve error message

lesteve committed:
[MRG] DOC good first issue and help wanted labels (#9950)

qinhanmin2014 committed:
DOC Encourage contributors to use keywords to close issue automatically (#9954)

jnothman committed:
Duplicate import

Commit:
[MRG+1] DOC fix sign in GBRT mathematical formulation (#9885)

sam09 committed:
Fix LogisticRegressionCV default solver value in docstring (#9962)

qinhanmin2014 committed:
[MRG] DOC Fix missing link in kernel_ridge.py (#9966)

* DOC Fix missing link in kernel_ridge.py

* remove func

kevin-coder committed:
Python 2 fix for plot_stock_market.py error (#9965)

zorroblue committed:
DOC Fix documentation for KDDCup99 dataset (#9974)

TomAugspurger committed:
DOC: Fixed typo (#9977)

hristog committed:
Improve readability of outlier detection example. (#9973)

jnothman committed:
DOC Add what's new for 0.19.1 (#9983)

lesteve committed:
DOC add missing dot in docstring

impaktor committed:
DOC Fix three typos in manifold documentation (#9990)

jnothman committed:
DOC update news

thechargedneutron committed:
[MRG+1] Deprecate pooling_func unused parameter in AgglomerativeClustering (#9875)

qinhanmin2014 committed:
improve example plot_forest_iris.py (#9989)

gkevinyen5418 committed:
DOC Fix typo: x axis -> y axis (#9985)

jinkunw committed:
DOC Fix typo (#9996)

gustavodemari committed:
[MRG + 1] Fix negative inputs checking in mean_squared_log_error (#9968)

* fixes msle when the inputs is negative, resolves #9963

* adding some regression tests for msle metric

albertcthomas committed:
[MRG+1] Fix LOF and Isolation benchmarks (#9798)

jnothman committed:
[MRG] FIX bug in nested set_params usage (#9999)

glemaitre committed:
[MRG + 1] ENH add check_inverse in FunctionTransformer (#9399)

* EHN add check_inverse in FunctionTransformer

* Add whats new entry and short narrative doc

* Sparse support

* better handle sparse data

* Address andreas comments

* PEP8

* Absolute tolerance default

* DOC fix docstring

* Remove random state and make check_inverse deterministic

* FIX remove random_state from init

* PEP8

* DOC motivation for the inverse

* make check_inverse=True default with a warning

* PEP8

* FIX get back X from check_array

* Andread comments

* Update whats new

* remove blank line

* joel s comments

* no check if one of forward or inverse not provided

* DOC fixes and example of filterwarnings

* DOC fix warningfiltering

* DOC fix merge error git

gxyd committed:
[MRG+1] correct comparison in GaussianNB for 'priors' (#10005)

liamge committed:
MAINT Remove redundancy in #9552 (#9573)

albertcthomas committed:
[MRG + 1] Labels of clustering should start at 0 or -1 if noise (#10015)

* test labels of clustering should start at 0 or -1 if noise

* take into account agramfort's comment

* fix test

srajanpaliwal committed:
[MRG] Fix LogisticRegression see also should include LogisticRegressionCV(#9995) (#10022)

gxyd committed:
[MRG+1] add changelog entry for fixed and merged PR #10005 issue #9633 (#10025)

* add changelog entry for fixed and merged PR #10005 issue #9633

* change name

* change PR number

muayyad-alsadi committed:
fixes #10031: fix attribute name and shape in documentation (#10033)

nnadeau committed:
[MRG+1] `MLPRegressor` quits fitting too soon due to `self._no_improvement_count` (#9457)

VinodKumarLogan committed:
[MRG+1] Remove sklearn.utils.testing._assert_all_close (#10032)

rth committed:
DOC Fix a few typos (#10038)

maskani-moh committed:
DOC Add references for multiclass balanced-accuracy definitions (#9982)

nsuh committed:
fix typo in docs - modules/model_persistence (#10047)

zorroblue committed:
[MRG+1] Added tests for parameter checks in GradientBoostingRegressor (#10013)

bodhwani committed:
[MRG] DOC Fix default learning_rate in SGDRegressor docstring (#10018)

* [MRG] Fix learning_rate in SGDRegressor docstring(#10012)

* Update SGDRegressor's learning_rate [default]

* Removed pep8 error

* Restored blank line

* Resolved pep8 error

CoderPat committed:
[MRG] Add check for n_components in pca (#10042)

qinhanmin2014 committed:
Fix PEP8 error for #10042

amueller committed:
make warning look way nicer in SGDClassifier (#10050)

qinhanmin2014 committed:
MAINT remove duplicate import in test_pca.py (#10061)

bodhwani committed:
[MRG] Added base.is_classifier/is_regressor in docs (#10062)

agramfort committed:
DOC Comparison plot for anomaly detection methods. (#10004)

jnothman committed:
DOC Note on _contributors.rst as its presence is now clearer

nzw0301 committed:
Fix links to Hoffman's onlineldavb code (#10070)

sergulaydore committed:
[MRG+2] faster way of computing means across each group (#10020)

lesteve committed:
TRAVIS install flake8 3.5 from pip (#10085)

jnothman committed:
DOC fix comment syntax

qinhanmin2014 committed:
DOC Fix some dead links in what's new 0.20 (#10104)

qinhanmin2014 committed:
[MRG] Ensure that ROC curve starts at (0, 0) (#10093)

oleksandr-pavlyk committed:
MAINT: only call clock() if verbosity level warrants it (#10091)

Put calls to `clock()` inside conditional statements.

This helps combat thread contention when executing TSNE in sklearn, compiled with icc.

Running TSNE on MNIST dataset (training + validation + test) of 70_000 hand-written images 28 by 28 pixels each, time of TSNE drops from 92 seconds to 81 seconds from this change alone.

jdoepfert committed:
[MRG + 1] Fix BayesianRidge() and ARDRegression() for constant target vectors  (#10095)

* add test for issue #10092

* add comment to test

* split into two tests

* add tests for scores, alpha and beta

* adapt tests: n_samples != n_features

* add test when no intercept is fitted

* add handling of constant target vector when intercept is fitted

* fix typo in comments

* fix format issues

* replace original fix with simpler fix

* add comment

* increase upper boundary for test

* increase upper boundary for test

* merge tests for ARDRegression and BayesianRidge

* use random state in tests

* decrease upper bound for std

* replace np.spacing(1) -> np.finfo(np.float64).eps

qinhanmin2014 committed:
DOC Fix dead links in SGD (#10109)

christianbraune79 committed:
FIX make_circles() now works with odd number of samples, test added (#10045)

qinhanmin2014 committed:
DOC Fix broken link in adjusted_mutual_info_score (#10123)

jnothman committed:
DOC Add Examples heading

jnothman committed:
CI temporarily use numpydoc master when building dev docs (#10066)

jnothman committed:
DOC Fix markup in docstring

jnothman committed:
DOC Correct attribute name

pstjohn committed:
ENH adding sample weights for BayesianRidge (#10112)

GKjohns committed:
TST Check estimator pairwise (#9701)

FarahSaeed committed:
[MRG] DOC Replacing "the scikit" with "scikit-learn"  (#10126)

siftikha committed:
[MRG] move flake8 options from flake8_diff.sh to setup.cfg (#10080)

Also add examples/.flake8 for examples specific flake8 configuration

lesteve committed:
Fix np.set_printoptions argument change in numpy 1.14.dev (#10132)

https://github.com/numpy/numpy/pull/9332/files change sign='legacy' to legacy=True in the np.set_printoptions arguments.

lesteve committed:
[MRG] DOC add documentation about Travis cron job (#10124)

qinhanmin2014 committed:
DOC remove broken link in faq.rst (#10135)

and related changes

lesteve committed:
Turn off codecov comments (#10146)

maskani-moh committed:
Balanced accuracy doc - 2 (#10040)

* Add references for multiclass balanced-accuracy

* Add precision not implemented

* Add another reference

* Adjust note

* Add succint definitions

* Add macro-average recall implementation

* Add class balance accuracy reference and definition

* Fix typo

* Add precision/recall comparison

* Fix typo

* Make descriptions less verbose and merge definitions

* Add reference for averaging accuracy

* Move macro-average recall example to corresponding section

siftikha committed:
[MRG+1] fix for erroneous max_iter and tol warnings for SGDClassifier when using partial_fit (#10053)

* partial fit warnings disabled

* partial fit warnings disabled for regressor

* style improved

* tests added

* pycodestyle passing

* rejiggered format

* fixed style issues

lesteve committed:
Make scipy.sparse.csgraph imports explicit (#10145)

This can cause problems in scipy 1.0

lesteve committed:
DOC fix link to SAG article

qinhanmin2014 committed:
DOC Fix broken link in ROC example (#10151)

jnothman committed:
DOC document some useful canned responses (#10131)

lesteve committed:
DOC improve titles for saved replies

TomDLT committed:
Fix underflow in nmf._beta_divergence (#10142)

lesteve committed:
[MRG+1] Remove nose from CIs and documentation (#9840)

agramfort committed:
Fix omp non normalize (#10071)

* fix OMP when columns of X are not normalized

qinhanmin2014 committed:
avoid subplot overlapping in plot_regression.py (#10159)

adamgreenhall committed:
[MRG+1] add groups support to RFECV (#9656)

anguswilliams91 committed:
[MRG+1] DOC Fix utils.resample and utils.shuffle docstrings: 'views' -> 'copies' (#10162)

lesteve committed:
MAINT remove _named_check (#10160)

which was a nose-specific thing to have good names for tests using yield

lesteve committed:
MAINT follow numpy dev change in np.set_printoptions argument

jnothman committed:
[MRG+1] MAINT Use magic to list documentation versions (#9841)

lesteve committed:
MAINT only build versions list on master branch

rth commented:
I restarted Travis CI and most of the builds now pass (so it was an intermittent failure), except for the PY2.7 build which fails with,
```
ImportError: Your installation of Numerical Python (NumPy) 1.6.1 is out-of-date.
scikit-learn requires NumPy >= 1.8.2.
```
I think this is something to do with the rebase. Could you rebase again please: that should fix it. Also, the current Circle CI error about failed download should go away if you re-start that build (or push any other commit or rebase).

rth commented on sklearn/metrics/classification.py:
```diff
@@ -179,6 +179,80 @@ def accuracy_score(y_true, y_pred, normalize=True, sample_weight=None):
+    result may be incorrect if one of those classes falls at the threshold, as
+    one class must be chosen to be the nth class and the class chosen may not
+    be the correct one.
+    Examples
```
line break missing above

rth commented on sklearn/metrics/classification.py:
```diff
@@ -179,6 +179,80 @@ def accuracy_score(y_true, y_pred, normalize=True, sample_weight=None):
+    1.0
+    >>> top_n_accuracy_score(y_true, y_pred, n=2, normalize=False)
+    3
+    """
```
It might be useful to add some reference papers / textbooks in the docstring if possible.

rth commented on sklearn/metrics/classification.py:
```diff
@@ -179,6 +179,80 @@ def accuracy_score(y_true, y_pred, normalize=True, sample_weight=None):
+    In cases where two or more classes are assigned equal likelihood, the
+    result may be incorrect if one of those classes falls at the threshold, as
+    one class must be chosen to be the nth class and the class chosen may not
+    be the correct one.
```
Incorrect - as in slightly inaccurate or completely wrong? Is there a way around it?

jwjohnson314 replied:
Incorrect. Consider a single observation whose target is class 1 (out of 5). Suppose the predicted probabilities are y_pred = [0.4, 0.2, 0.1, 0.2, 0.1]. The top 3 predictions are 0, 1, and 3, so the top-3 accuracy is 1.0 on this observation. The top-2 predictions are 0 and 3 (np.argsort returns indices containing equal values in ascending order, and this metric takes the tail of the argsort list), so the top-2 accuracy is 0 on this observation: even though classes 1 and 3 are assigned the same probabilities, class 1 doesn't make the cut. I could replace the disclaimer with logic to address this - in, tensorflow, for instance, any class assigned probability equal to a probability that made the top-n list to be in the top n. With that change the top-2 prediction in the example described would be the same as the top-3 accuracy. Let me know what you think.

rth commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -101,6 +102,21 @@ def make_prediction(dataset=None, binary=False):
+    # edge case, perfect accuracy
+    assert_equal(top_n_accuracy_score(y_true, pred_proba, n=3), 1.0)
+    # edge case. Should be the same as accuracy_score, but 'probabilities'
+    # for SVM don't always align with class predictions. This happens with
```
Do you mean that the class with the highest probability is not necessarily the one getting predicted in `SVC`? Do you have a reference for this? 

It looks like the probabilities from `make_prediction` are never used in `sklearn/metrics/tests/test_classification.py` although they are used in `test_ranking.py`. I'm not sure what was the original motivation for using SVC in this benchmark function (last modified 3 years ago), but if what you are saying is confirmed I don't think that attempting to fix it's output is the right approach. IMO it would be better to fix it in `make_prediction` e.g. by adding an additional parameter allowing to use a different classifier that doesn't have this issues.

amueller replied:
Why use any model at all and not just create some probabilities?

amueller replied:
Also, there should be some non-trivial tests that are easy to check "by hand"

jwjohnson314 replied:
I was thinking specifically of the sklearn implementation with Platt scaling. The issue is described in the docs [here](http://scikit-learn.org/stable/modules/svm.html) in section 1.4.1.2 for a binary case. @amueller no particular reason to use a model; just trying to take advantage of what was already there - I'll make some simple cases that can be verified by hand.

rth reviewed (changes requested):
Thanks for the PR. A few comments are below,
 * you should add this score to common tests in [`sklearn/metrics/tests/test_common.py`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/tests/test_common.py)

 * with respect to the naming, is "top n accuracy" more standard than "top k accuracy"? Just wondering about  the consistency of the naming with respect to e.g. "precision at k" (https://github.com/scikit-learn/scikit-learn/issues/7343) or k-nearest neighbours (though it's much less related)

qinhanmin2014 committed:
DOC previous versions -> all available versions (#10178)

jorisvandenbossche committed:
ENH: new CategoricalEncoder class (#9151)

abjer committed:
FIX Return nan in RadiusNeighborsRegressor for empty neighbor set (#9655)

* Fix #9654

NicolasHug committed:
Fix typo in ShuffleSplit docstring and effective_n_jobs (#10182)

MarkIwanchyshyn committed:
DOC Fix Agglomerative children_ documentation (#10184)

reiinakano committed:
[MRG+1] classification_report: raise error if labels is None and target_names size is not equal to present classes (#9867)

jdoepfert committed:
[MRG + 1] FIX Calculation of standard deviation of predictions in ARDRegression (#10153)

TomDLT committed:
reference issue number in whats_new

jorisvandenbossche committed:
CLN: remove nose fixtures in docs (#10187)

211217613 committed:
DOC Change docstring format to match other docstrings (#10190)

The wording was a little funny, also I changed it to match the rest
of the docstrings in other files

amueller committed:
[MRG] fix command for make test-coverage (#10188)

patrick1011 committed:
[MRG] Added missing sag_solver documentation (#10172)

qinhanmin2014 committed:
DOC change doc link about testing from stable (nose) to dev (pytest)

We have changed to pytest but the stable doc is about nose.
Already gain approval from @lesteve in #9840.
Nothing to check so I push directly.

varunagrawal committed:
[MRG + 1] Fix for image.extract_patches_2d when # of patches requested > all possible patches  (#10101)

trevorstephens committed:
MAINT Update libatlas version in advanced installation instructions (#10204)

FarahSaeed committed:
TEST caught deprecation warning for min_impurity_split  (#10202)

jnothman committed:
DOC clearer wording in QuantileTransformer user guide

lesteve committed:
MAINT use Python 3.4 in first conda build (#10193)

fenilsuchak committed:
[MRG+1] Error raised when presort in gradient_boosting.py and tree.py is not a boolean (#10110)

mfeurer committed:
DOC add docs about sgd and warmstarting (#10011) (#10087)

In particular, clarify for SGDClassifier, SGDRegressor,
PassiveAggressiveClassifier and PassiveAggressiveRegressor how
the attribute warm_start and the methods `fit` and `partial_fit`
interact.

Fixes #10011

rth committed:
[MRG+1] Support for 64 bit sparse array indices in text vectorizers (#9147)

Support new scipy sparse array indices, which can now be > 2^31 (< 2^63).
This is needed for very large training sets.
Feature indices (based on the number of distinct features), are unlikely to need 4 bytes per value, however.

cbrummitt committed:
[MRG + 1] DOC Fix cardinality vs. l0 norm in multilabel ranking user guide (#10219)

jwjohnson314 commented:
Thanks for the feedback. I have very little bandwidth for the next week or so but will get back to this as soon as I can.

glemaitre committed:
[MRG+1] FIX convert to CSR in decision_function (#10102)

christianbraune79 committed:
[MRG+1] Warning in KMeans if too few clusters were found (#10099)

* give warning if number of clusters found is smaller than number of clusters asked for

* unnecessary comment

* tests commented, pep8, ConvergenceWarning instead of RuntimeWarning

* lesteve's comment

FarahSaeed committed:
MAINT Removing 'random_state = 0' while intializing OneClassSVM (#10241)

and some PEP8 fixes

phausamann committed:
DOC Add sklearn-xarray package to related projects (#10243)

jnothman committed:
Improve error message when list_versions fails on http request

mabelvj committed:
FIX export_graphviz should work with a sklearn.tree._tree.Tree (#10234)

maskani-moh committed:
[MRG+1] Take over PR #7647 - Add a "filename" attribute to datasets that have a CSV file (#9101)

* add filename attribute for load_iris

* add filename attribute for load_boston

* add filename attribute for load_linnerud

lesteve committed:
CI Run make test-doc on scipy-dev-wheels build (#10246)

rajendraarora16 committed:
DOC Fixing "dictionary" spelling (#10255)

frsi committed:
DOC Fixed BinaryTree Documentation Examples (issue #10199) (#10205)

rajendraarora16 committed:
DOC Improve comment in plot_classifier_comparison.py (#10256)

chang committed:
Feature: Implement PowerTransformer (#10210)

jnothman committed:
[MRG + 1] DOC Update doc/README.md (#10225)

gxyd committed:
DOC fix docstring Gradient Boosting (#10262)

frexvahi committed:
DOC: 'all cannot' -> 'not all can' (#10263)

tttthomasssss committed:
[MRG+1] Fixes #9199 - Fitting a NearestNeighbors model fails with sparse input and a callable as metric (#9579)

adelr committed:
[MRG+1] DOC Mention MKL multithreading in FAQ

jbearer committed:
MAINT Hide train_test_split from nose test discovery. (#10223)

gxyd committed:
[MRG+1] Contingency matrix doc (#10220)

* Update example to clarify difference with Confusion Matrix, added link to Confusion Matrix documentation

* fixed 80 character limit issue

* changed indentation for automated test

* fixed formatting

* fixed formatting

* fixed formatting

* address requested changes in Contingency Matrix docs

* basic formatting

* use less technical and correct definition of contingency matrix

* arrange in alphabetical order

* add 'advantages' and 'disadvantages' of contingency matrix

* use better 'str' labels for example

* address the language of documentation

* add `ref` in documentation

* add a better 'advantage'

* fix

* address review

* small change

* add full stop.

mirca committed:
DOC Clarify notation for logistic regression objective function (#10269)

jnothman committed:
DOC Add NORMALIZE_WHITESPACE for numpy master

maskani-moh committed:
DOC Add entry in change log in whats_new.rst (#10285)

thechargedneutron committed:
FEA Add a new class RegressorChain similar to ClassifierChain (#9257)

agramfort committed:
Deprecate lar positive (#10248)

* deprecate positive=True option in lasso_lars with method='lar'

amueller committed:
API deprecate *SearchCV iid parameter (#9379)

jnothman commented:
Thanks for the heads up, @jwjohnson314. Please let us know if you will not be able to complete this.

alexlenail committed:
DOC: add graph spectral clustering doc to user guide (fix #9481) (#9811)

NicolasHug committed:
DOC Add scikit-surprise as a related project (#8148)

jnothman committed:
DOC fix doctest

jnothman committed:
DOC fix NameError in doctest

amueller committed:
make SGDClassifier deprecation warnings nice again (#10286)

caioaao committed:
Remove hanging comma (#10294)

lesteve committed:
MAINT install flake8 from conda. (#10293)

flake8 3.5 is now available from conda

jrbourbeau committed:
DOC Updates contributing.rst with doc build instructions (#10272)

thechargedneutron committed:
DOC Corrects links to doc of ClassifierChain and RegressorChain (#10288)

lesteve committed:
MAINT remove unicode character from URL for doc linkcheck friendliness (#10300)

herilalaina committed:
[MRG+1] - No private setting in __init__ apart from parameters (#9450)

lesteve committed:
[MRG] DOC run three more examples that use sys.argv (#10301)

amueller committed:
DOC add scikit-optimize to related projects (#10296)

glemaitre committed:
FEA TransformedTargetRegressor (#9041)

Add a new meta-regressor which transforms y for training.

kno10 committed:
DOC Fix equations for mutual_information. (#10292)

tttthomasssss committed:
DOC Added entry for #9579 to Whats New (#10312)

jnothman committed:
[MRG] TST add test for silhouette from its original publication (#10298)

chang committed:
DOC Update docs for assertion functions in sklearn.utils.testing (#10317)

chang committed:
[MRG+1] ENH Add standardize flag to PowerTransformer (#10316)

tuliocasagrande committed:
Improve readability of LOF example (#10322)

TsingQAQ committed:
DOC replace 'maximize' by 'minimize' in gpr.py (#10327)

vrishank97 committed:
DOC Add paired_distances in pairwise_distances docstring (#10304)

jmargeta committed:
Fix spectral embedding implementation (#9062)

lesteve committed:
DOC fix rst excessive bullet point indentation (#10335)

gxyd committed:
ENH Make GradientBoostingClassifier error message more informative (#10207)

jwjohnson314 commented:
@rth you are right, top-k accuracy is more conventional and more consistent with the other metrics you mention.

Commit:
TST: replaced top_k_accuracy_score test with easily verifable by hand
case. Changed name from top_n_accuracy_score to top_k_accuracy_score.
Fixed minor typos and edited docstring.

Commit:
resolved merge conflict

jnothman commented:
This pull request introduces 4 alerts and fixes 21 - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-4579ccc97c133aa92317196354cf9c127392a1a8)

**new alerts:**

* 1 for __eq__ not overridden when adding attributes
* 1 for Syntax error
* 1 for Explicit export is not defined
* 1 for Mismatch between signature and use of an overridden method

**fixed alerts:**

* 4 for Use of the return value of a procedure
* 4 for Missing call to __init__ during object initialization
* 3 for Variable defined multiple times
* 2 for Comparison using is when operands support __eq__
* 1 for Too few arguments in formatting call
* 1 for Wrong name for an argument in a call
* 1 for Inconsistent equality and inequality
* 1 for Conflicting attributes in base classes
* 1 for Special method has incorrect signature
* 1 for Unreachable code
* 1 for Redundant assignment
* 1 for Potentially uninitialized local variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

amueller commented:
looks like a rebase went wrong or something.

Commit:
fixed missed namechanges

jnothman commented:
This pull request introduces 3 alerts and fixes 21 - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-9128dc2861765aec06860f3a01cefcd0af4f4f8f)

**new alerts:**

* 1 for __eq__ not overridden when adding attributes
* 1 for Syntax error
* 1 for Mismatch between signature and use of an overridden method

**fixed alerts:**

* 4 for Use of the return value of a procedure
* 4 for Missing call to __init__ during object initialization
* 3 for Variable defined multiple times
* 2 for Comparison using is when operands support __eq__
* 1 for Too few arguments in formatting call
* 1 for Wrong name for an argument in a call
* 1 for Inconsistent equality and inequality
* 1 for Conflicting attributes in base classes
* 1 for Special method has incorrect signature
* 1 for Unreachable code
* 1 for Redundant assignment
* 1 for Potentially uninitialized local variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

rth commented:
@jwjohnson314 Yes, it looks like the rebase didn't work out. To fix this you could, rebase your master to the latest version, create a new branch from it and then cherry-pick the few commits corresponding to this PR. Feel free to create a new PR from that branch if it's easier. Let me know if you need more help with that.

jnothman commented:
Or rather than cherry-picking commits, just patch the files you want and make a new commit and PR.

rth commented:
Continued in https://github.com/scikit-learn/scikit-learn/pull/10488 (also no reason to keep this one open since the rebase went wrong).
